File: src/main/java/org/kite9/framework/logging/Table.java
package org.kite9.framework.logging;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Tabular displayer
 * 
 */
public class Table {

	List<Integer> widths = new ArrayList<Integer>();

	List<String[]> rows = new ArrayList<String[]>();

	public void addRow(Object... items) {
		List<String> longRow = new ArrayList<String>();

		for (Object object : items) {
			if (object==null) {
				longRow.add("null");
			} else if (object.getClass().isArray()) {
				int length = Array.getLength(object);

				for (int i = 0; i < length; i++) {
					Object item = Array.get(object, i);
					longRow.add(item.toString());
				}
			} else if (object instanceof Collection<?>) {
				Collection<?> c = (Collection<?>) object;
				for (Object object2 : c) {
					longRow.add(object2.toString());
				}
			} else {
				longRow.add(object.toString());
			}
		}

		addArrayRow((String[]) longRow.toArray(new String[longRow.size()]));

	}
	
	public void removeLastRow() {
		rows.remove(rows.size()-1);
	}

	public void addArrayRow(String[] row) {
		rows.add(row);

		int col = 0;
		for (String string : row) {
			if (widths.size() > col) {
				widths.set(col, widths.get(col) > string.length() ? widths.get(col) : string.length());
			} else {
				widths.add(string.length());
			}
			col++;
		}
	}
	
	public void addObjectRow(Object[] row) {
		String[] items = new String[row.length];
		for (int i = 0; i < items.length; i++) {
			items[i] = row[i].toString();
		}
		addArrayRow(items);
	}
	
	public void addDoubleRow(double[] row) {
		String[] items = new String[row.length];
		for (int i = 0; i < items.length; i++) {
			items[i] = Double.toString(row[i]);
		}
		addArrayRow(items);
	}

	public void display(StringBuffer sb) {
		for (String[] row : rows) {
			int colno = 0;
			for (String col : row) {
				int width = widths.get(colno) + 1;
				sb.append(col);
				for (int i = col.length(); i < width; i++) {
					sb.append(" ");
				}

				colno++;
			}
			sb.append("\n");
		}
	}
	
}

File: src/main/kotlin/org/kite9/diagram/logging/Table.kt
package org.kite9.diagram.logging

/**
 * Tabular displayer
 *
 */
class Table {

    var widths: MutableList<Int> = ArrayList()
    var rows: MutableList<Array<String?>> = ArrayList()

    fun addRow(vararg items: Any?) {
        val longRow: MutableList<String?> = ArrayList()
        for (o in items) {
            if (o == null) {
                longRow.add("null")
            } else if (o is Array<*>) {
                val length = o.size
                for (i in 0 until length) {
                    val item = o[i]
                    longRow.add(item.toString())
                }
            } else if (o is Collection<*>) {
                for (object2 in o) {
                    longRow.add(object2?.toString() ?: "")
                }
            } else {
                longRow.add(o.toString())
            }
        }
        addArrayRow(longRow.toTypedArray())
    }

    fun removeLastRow() {
        rows.removeAt(rows.size - 1)
    }

    fun addArrayRow(row: Array<String?>) {
        rows.add(row)
        var col = 0
        for (string in row) {
            if (widths.size > col) {
                widths[col] = if (widths[col] > string!!.length) widths[col] else string.length
            } else {
                widths.add(string!!.length)
            }
            col++
        }
    }

    fun <T> addObjectRow(row: Array<T>) {
        val items = arrayOfNulls<String>(row.size)
        for (i in items.indices) {
            items[i] = row[i].toString()
        }
        addArrayRow(items)
    }

    fun addDoubleRow(row: DoubleArray) {
        val items = arrayOfNulls<String>(row.size)
        for (i in items.indices) {
            items[i] = row[i].toString()
        }
        addArrayRow(items)
    }

    fun display(sb: StringBuilder) {
        for (row in rows) {
            var colno = 0
            for (col in row) {
                val width = widths[colno] + 1
                sb.append(col)
                for (i in col!!.length until width) {
                    sb.append(" ")
                }
                colno++
            }
            sb.append("\n")
        }
    }
}